File: SenLib/Sen2/FileFixes/scripts_scena_dat_us_t0010_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen2.FileFixes {
	class scripts_scena_dat_us_t0010_dat : FileMod {
		public string GetDescription() {
			return "Fix extra word in Thors (bottom floor).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x7f1a94208801acaful, 0x1a3ba2fd7cba4964ul, 0x44904a1bu));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var patcher = new SenScriptPatcher(bin);

			// during the our days in the army -> during our days in the army
			patcher.RemovePartialCommand(0x4201, 0x11b, 0x42dc, 0x4);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0010.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x7f1a94208801acaful, 0x1a3ba2fd7cba4964ul, 0x44904a1bu));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0010.dat", file) };
		}
	}
}

File: native/sen2/file_fixes/scripts_scena_dat_us_t0010_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "util/hash/sha1.h"

namespace SenLib::Sen2::FileFixes::scripts_scena_dat_us_t0010_dat {
std::string_view GetDescription() {
    return "Fix extra word in Thors (bottom floor).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_us/t0010.dat",
            31169,
            HyoutaUtils::Hash::SHA1FromHexString("7f1a94208801acaf1a3ba2fd7cba496444904a1b"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        SenScriptPatcher patcher(bin);

        // during the our days in the army -> during our days in the army
        // This is during the very first chance you get to walk around in the Finale, in Vandyck's
        // office.
        patcher.RemovePartialCommand(0x4201, 0x11b, 0x42dc, 0x4);

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen2::FileFixes::scripts_scena_dat_us_t0010_dat
